OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs

```diff
@@ -340,22 +372,17 @@ void CancelProductionInner(string itemName)
 		{
 			var lastIndex = queue.FindLastIndex(a => a.Item == itemName);
 
-			if (lastIndex > 0)
-				queue.RemoveAt(lastIndex);
-			else if (lastIndex == 0)
+			if (lastIndex != -1)
 			{
-				var item = queue[0];
-
-				// Refund what has been paid
+				var item = queue[lastIndex];
 				playerResources.GiveCash(item.TotalCost - item.RemainingCost);
-				FinishProduction();
+				queue.RemoveAt(lastIndex);
 			}
 		}
 
-		public void FinishProduction()
+		public void FinishProduction(string itemName)
 		{
-			if (queue.Count != 0)
-				queue.RemoveAt(0);
+			queue.RemoveAt(queue.FindIndex(a => a.Item == itemName));
```

Comment:
is there a reason why you don't pass the ProductionItem itself to the FinishProduction method instead of its name?

Reply:
Nope, i just stupidly replaced the RemoveAt() method call. Ill change that.

Reply:
If it is too much work or impossible to change it I wouldn't worry about it too much.
I am mainly wondering if this will ever be expanded to allow for full synchronous production, aka multiple units of the same type in the same queue being created at the same time, if we ever did that it would probably be a better idea to have the item there, not sure it matters in your use case.

Reply:
The major advantage by multiple different units is you can push an harvester in between without the need to cancel everything else, or such things. If a single unit would be build multiple times in parallel, you wont have any advantage. More you will have a disadvantage if you need units quickly, you need to build one after another manualy.